In [1]:
import math
from decimal import *
import numpy as np # linear algebra
import pandas as pd # read_csv and such
from geopy.distance import vincenty # calculate distance between locations
from io import StringIO # convert strings to buffers or something like that.
import os # for listing files in directory

pd.options.display.max_columns = 999

In [2]:
# the paths to our flight related data sets
airlines_path = '../data/flight_delays_2015/airlines.csv'
airports_path = '../data/flight_delays_2015/airports.csv'
flights_path  = '../data/flight_delays_2015/flights.csv'
# weather related paths
weather_path = '../data/weather/2015.csv'
stations_path  = '../data/weather/stations.csv'
weather_files_path = '../data/weather/station_data/'
weather_merged_path = '../data/weather/weather_merged.csv'

# modified data sets for output
airports_modified_path = '../data/flight_delays_2015/airports_modified.csv'
flights_weather_path = '../data/flight_delays_2015/flights_weather.csv'

In [3]:
dtypes = {
    'ORIGIN_AIRPORT': 'str', 
    'DESTINATION_AIRPORT': 'str', 
    'IATA_CODE_x': 'str', 
    'origin_weather_station': 'str', 
    'IATA_CODE_y': 'str', 
    'destination_weather_station': 'str', 
    'OR_MAX': 'str', 
    'OR_MIN': 'str', 
    'OR_PRCP': 'str', 
    'DES_MAX': 'str', 
    'DES_MIN': 'str', 
    'DES_PRCP': 'str', 
    'OR_FRSHTT': 'str', 
    'DES_FRSHTT': 'str'
}
fw_df = pd.read_csv(flights_weather_path, dtype=dtypes, parse_dates=['DATE'])
fw_df.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      1    1            4      AS             98      N407AS   
1  2015      1    1            4      AS            108      N309AS   
2  2015      1    1            4      DL           1560      N3743H   
3  2015      1    1            4      AS            136      N431AS   
4  2015      1    1            4      AS            134      N464AS   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  DEPARTURE_TIME  \
0            ANC                 SEA                    5          2354.0   
1            ANC                 SEA                   45            41.0   
2            ANC                 SEA                   45            31.0   
3            ANC                 SEA                  135             NaN   
4            ANC                 SEA                  155           140.0   

   DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  \
0            -11.0      21.0        15.0           205.0         194.0   
1             -4.0      17.0        58.0           204.0         194.0   
2            -14.0      25.0        56.0           210.0         200.0   
3              NaN       NaN         NaN           205.0           NaN   
4            -15.0      17.0       157.0           218.0         198.0   

   AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  \
0     169.0      1448      404.0      4.0                430         408.0   
1     173.0      1448      451.0      4.0                509         455.0   
2     171.0      1448      447.0      4.0                515         451.0   
3       NaN      1448        NaN      NaN                600           NaN   
4     170.0      1448      547.0     11.0                633         558.0   

   ARRIVAL_DELAY  DIVERTED  CANCELLED CANCELLATION_REASON  AIR_SYSTEM_DELAY  \
0          -22.0         0          0                 NaN               NaN   
1          -14.0         0          0                 NaN               NaN   
2          -24.0         0          0                 NaN               NaN   
3            NaN         0          1                   A               NaN   
4          -35.0         0          0                 NaN               NaN   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  \
0             NaN            NaN                  NaN            NaN   
1             NaN            NaN                  NaN            NaN   
2             NaN            NaN                  NaN            NaN   
3             NaN            NaN                  NaN            NaN   
4             NaN            NaN                  NaN            NaN   

        DATE IATA_CODE_x origin_weather_station IATA_CODE_y  \
0 2015-01-01         ANC           702725-26491         SEA   
1 2015-01-01         ANC           702725-26491         SEA   
2 2015-01-01         ANC           702725-26491         SEA   
3 2015-01-01         ANC           702725-26491         SEA   
4 2015-01-01         ANC           702725-26491         SEA   

  destination_weather_station  OR_TEMP  OR_COUNT_TEMP  OR_DEWP  OR_COUNT_DEWP  \
0                727930-24233     35.1             24     30.7             24   
1                727930-24233     35.1             24     30.7             24   
2                727930-24233     35.1             24     30.7             24   
3                727930-24233     35.1             24     30.7             24   
4                727930-24233     35.1             24     30.7             24   

   OR_SLP  OR_COUNT_SLP  OR_STP  OR_COUNT_STP  OR_VISIB  OR_COUNT_VISIB  \
0  1020.1            14  1015.2            24       7.5              24   
1  1020.1            14  1015.2            24       7.5              24   
2  1020.1            14  1015.2            24       7.5              24   
3  1020.1            14  1015.2            24       7.5              24   
4  1020.1            14  1015.2            24     

Convert some of these columns to numeric data and remove values that indicated as missing.

In [4]:
fw_df['OR_PRCP'] = pd.to_numeric(fw_df['OR_PRCP'].str.replace('A|B|C|D|E|F|G|H|I', ''))
fw_df['OR_MIN'] = pd.to_numeric(fw_df['OR_MIN'].str.replace('*', ''))
fw_df['OR_MAX'] = pd.to_numeric(fw_df['OR_MAX'].str.replace('*', ''))

In [5]:
fw_df.loc[fw_df['OR_TEMP'] == 9999.9, 'OR_TEMP'] = np.NaN
fw_df.loc[fw_df['OR_MIN'] == 9999.9, 'OR_MIN'] = np.NaN
fw_df.loc[fw_df['OR_MAX'] == 9999.9, 'OR_MAX'] = np.NaN
fw_df.loc[fw_df['OR_WDSP'] == 999.9, 'OR_WDSP'] = np.NaN
fw_df.loc[fw_df['OR_MXSPD'] == 999.9, 'OR_MXSPD'] = np.NaN
fw_df.loc[fw_df['OR_GUST'] == 999.9, 'OR_GUST'] = np.NaN
fw_df.loc[fw_df['OR_SNDP'] == 999.9, 'OR_SNDP'] = np.NaN
fw_df.loc[fw_df['OR_PRCP'] == 99.99, 'OR_PRCP'] = np.NaN

### Create datafames that contain where flights are only weather delayed at origin or at destination
This will not contain all delays that fit these categories but those that we can be fairly certain fit a category.

Find destination only delays by checking where the flight departed on time, but arrived late and has a weather delay.

In [6]:
dest_delay_df = fw_df[(fw_df['WEATHER_DELAY'] > 0) & (fw_df['DEPARTURE_DELAY'] < 0) & (fw_df['ARRIVAL_DELAY'] > 0)]
dest_delay_df.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
8318   2015      4   15            3      VX            781      N524VA   
9549   2015      6    5            5      VX            781      N837VA   
11037  2015      8    4            2      VX            781      N852VA   
11822  2015      9    4            5      VX            781      N524VA   
16164  2015      4   14            2      VX            740      N521VA   

      ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  DEPARTURE_TIME  \
8318             LAX                 SEA                  730           729.0   
9549             LAX                 SEA                  745           743.0   
11037            LAX                 SEA                  745           744.0   
11822            LAX                 SEA                  800           759.0   
16164            SFO                 SEA                  800           755.0   

       DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  \
8318              -1.0       9.0       738.0           155.0         177.0   
9549              -2.0       9.0       752.0           150.0         168.0   
11037             -1.0      10.0       754.0           150.0         167.0   
11822             -1.0      16.0       815.0           150.0         176.0   
16164             -5.0      13.0       808.0           120.0         140.0   

       AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  SCHEDULED_ARRIVAL  \
8318      157.0       954     1015.0     11.0               1005   
9549      146.0       954     1018.0     13.0               1015   
11037     145.0       954     1019.0     12.0               1015   
11822     147.0       954     1042.0     13.0               1030   
16164     118.0       679     1006.0      9.0               1000   

       ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED CANCELLATION_REASON  \
8318         1026.0           21.0         0          0                 NaN   
9549         1031.0           16.0         0          0                 NaN   
11037        1031.0           16.0         0          0                 NaN   
11822        1055.0           25.0         0          0                 NaN   
16164        1015.0           15.0         0          0                 NaN   

       AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
8318                0.0             0.0            0.0                  0.0   
9549                0.0             0.0            0.0                  0.0   
11037               0.0             0.0            0.0                  0.0   
11822               0.0             0.0            0.0                  0.0   
16164               0.0             0.0            0.0                  0.0   

       WEATHER_DELAY       DATE IATA_CODE_x origin_weather_station  \
8318            21.0 2015-04-15         LAX           722950-23174   
9549            16.0 2015-06-05         LAX           722950-23174   
11037           16.0 2015-08-04         LAX           722950-23174   
11822           25.0 2015-09-04         LAX           722950-23174   
16164           15.0 2015-04-14         SFO           724940-23234   

      IATA_CODE_y destination_weather_station  OR_TEMP  OR_COUNT_TEMP  \
8318          SEA                727930-24233     63.7             24   
9549          SEA                727930-24233     62.7             24   
11037         SEA                727930-24233     72.2             24   
11822         SEA                727930-24233     69.8             24   
16164         SEA                727930-24233     55.8             24   

       OR_DEWP  OR_COUNT_DEWP  OR_SLP  OR_COUNT_SLP  OR_STP  OR_COUNT_STP  \
8318      48.2             24  1014.4            24  1002.6            24   
9549      55.3             24  1011.1            22   999.4            24   
11037     63.0             24  1012.1            24  1000.6            24   
11822     57.3             24  1011.0            23   999.5            24

Find origin delays by checking that there is a weather delay, the delay at arrival is no more than departure and there is a departure delay.

In [7]:
origin_delay_df = fw_df[(fw_df['WEATHER_DELAY'] > 0) & (fw_df['ARRIVAL_DELAY'] <= fw_df['DEPARTURE_DELAY']) & (fw_df['DEPARTURE_DELAY'] > 0)]
origin_delay_df.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
5580  2015     11   10            2      AS            108      N319AS   
5851  2015     11   28            6      DL            822      N372DA   
6900  2015      1   31            6      OO           5608      N789SK   
7674  2015      3   15            7      AS            449      N795AS   
7932  2015      3   28            6      OO           5608      N919SW   

     ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  DEPARTURE_TIME  \
5580            ANC                 SEA                   45           126.0   
5851            ANC                 SEA                   50           134.0   
6900            LAX                 SEA                  840           928.0   
7674            LAX                 SEA                 1350          1457.0   
7932            LAX                 SEA                  920          1021.0   

      DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  \
5580             41.0      15.0       141.0           201.0         200.0   
5851             44.0      12.0       146.0           205.0         201.0   
6900             48.0      10.0       938.0           164.0         160.0   
7674             67.0      20.0      1517.0           167.0         153.0   
7932             61.0      19.0      1040.0           175.0         156.0   

      AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  \
5580     179.0      1448      540.0      6.0                506         546.0   
5851     182.0      1448      548.0      7.0                515         555.0   
6900     143.0       954     1201.0      7.0               1124        1208.0   
7674     126.0       954     1723.0      7.0               1637        1730.0   
7932     132.0       954     1252.0      5.0               1215        1257.0   

      ARRIVAL_DELAY  DIVERTED  CANCELLED CANCELLATION_REASON  \
5580           40.0         0          0                 NaN   
5851           40.0         0          0                 NaN   
6900           44.0         0          0                 NaN   
7674           53.0         0          0                 NaN   
7932           42.0         0          0                 NaN   

      AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
5580               0.0             0.0            0.0                  0.0   
5851               0.0             0.0            0.0                  0.0   
6900               0.0             0.0            0.0                  9.0   
7674               0.0             0.0            0.0                  0.0   
7932               0.0             0.0            0.0                  0.0   

      WEATHER_DELAY       DATE IATA_CODE_x origin_weather_station IATA_CODE_y  \
5580           40.0 2015-11-10         ANC           702725-26491         SEA   
5851           40.0 2015-11-28         ANC           702725-26491         SEA   
6900           35.0 2015-01-31         LAX           722950-23174         SEA   
7674           53.0 2015-03-15         LAX           722950-23174         SEA   
7932           42.0 2015-03-28         LAX           722950-23174         SEA   

     destination_weather_station  OR_TEMP  OR_COUNT_TEMP  OR_DEWP  \
5580                727930-24233     31.1             24     27.7   
5851                727930-24233     35.5             24     31.1   
6900                727930-24233     60.0             24     55.1   
7674                727930-24233     74.9             24     40.9   
7932                727930-24233     64.0             24     56.5   

      OR_COUNT_DEWP  OR_SLP  OR_COUNT_SLP  OR_STP  OR_COUNT_STP  OR_VISIB  \
5580             24   995.6            17   990.9            24       8.8   
5851             24   997.1            24   992.4            24      10.0   
6900             24  1011.5            22   999.9            24       8.5   
7674             24  1016.1            24  1004.4            24   

In [8]:
both_delay_df = fw_df[(fw_df['WEATHER_DELAY'] > 0) & (fw_df['ARRIVAL_DELAY'] > fw_df['DEPARTURE_DELAY']) & (fw_df['DEPARTURE_DELAY'] > 0) & ~(fw_df['AIR_SYSTEM_DELAY'] > 0) & ~(fw_df['SECURITY_DELAY'] > 0) & ~(fw_df['AIRLINE_DELAY'] > 0) & ~(fw_df['LATE_AIRCRAFT_DELAY'] > 0)]
both_delay_df.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
136564  2015      9    3            4      HA             26      N390HA   
453853  2015     11   20            5      HA             21      N384HA   
466913  2015      9    3            4      HA            365      N486HA   
470493  2015      2   14            6      HA            119      N492HA   
474731  2015      9   15            2      HA            279      N490HA   

       ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  \
136564            HNL                 PDX                 1405   
453853            SEA                 HNL                  835   
466913            OGG                 HNL                 1400   
470493            KOA                 HNL                 1004   
474731            KOA                 HNL                 1746   

        DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  SCHEDULED_TIME  \
136564          1502.0             57.0      33.0      1535.0           330.0   
453853           843.0              8.0      31.0       914.0           380.0   
466913          1427.0             27.0       5.0      1432.0            37.0   
470493          1110.0             66.0       7.0      1117.0            44.0   
474731          1751.0              5.0      22.0      1813.0            44.0   

        ELAPSED_TIME  AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  \
136564         350.0     309.0      2603     2344.0      8.0   
453853         391.0     353.0      2677     1307.0      7.0   
466913          66.0      37.0       100     1509.0     24.0   
470493          56.0      35.0       163     1152.0     14.0   
474731          61.0      33.0       163     1846.0      6.0   

        SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
136564               2235        2352.0           77.0         0          0   
453853               1255        1314.0           19.0         0          0   
466913               1437        1533.0           56.0         0          0   
470493               1048        1206.0           78.0         0          0   
474731               1830        1852.0           22.0         0          0   

       CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
136564                 NaN               0.0             0.0            0.0   
453853                 NaN               0.0             0.0            0.0   
466913                 NaN               0.0             0.0            0.0   
470493                 NaN               0.0             0.0            0.0   
474731                 NaN               0.0             0.0            0.0   

        LATE_AIRCRAFT_DELAY  WEATHER_DELAY       DATE IATA_CODE_x  \
136564                  0.0           77.0 2015-09-03         HNL   
453853                  0.0           19.0 2015-11-20         SEA   
466913                  0.0           56.0 2015-09-03         OGG   
470493                  0.0           78.0 2015-02-14         KOA   
474731                  0.0           22.0 2015-09-15         KOA   

       origin_weather_station IATA_CODE_y destination_weather_station  \
136564           911820-22521         PDX                726980-24229   
453853           727930-24233         HNL                911820-22521   
466913           911900-22516         HNL                911820-22521   
470493           911975-21510         HNL                911820-22521   
474731           911975-21510         HNL                911820-22521   

        OR_TEMP  OR_COUNT_TEMP  OR_DEWP  OR_COUNT_DEWP  OR_SLP  OR_COUNT_SLP  \
136564     80.9             24     75.4             24  1011.0            22   
453853     39.1             24     33.1             24  1031.1            23   
466913     81.8             24     72.7             24  1012.2            24   
470493     74.9             24     68.0             24  1007.9            24   
474731     84.4             24     72.9             24  1012.0            24  

Since the dataframe with flights that have origin based weather delays are the only one that has significant number of entries, let's focus some basic analysis on these.

First, let's check the mean of the weather values relative to the mean of all the flights.

In [13]:
fw_df[['OR_TEMP', 'OR_MIN', 'OR_MAX', 'OR_WDSP', 'OR_MXSPD', 'OR_GUST', 'OR_PRCP', 'OR_SNDP']].mean()

OR_TEMP     61.333791
OR_MIN      51.190735
OR_MAX      70.381713
OR_WDSP      7.025933
OR_MXSPD    14.014208
OR_GUST     23.073898
OR_PRCP      0.111353
OR_SNDP      4.516190
dtype: float64

In [12]:
origin_delay_df[['OR_TEMP', 'OR_MIN', 'OR_MAX', 'OR_WDSP', 'OR_MXSPD', 'OR_GUST', 'OR_PRCP', 'OR_SNDP']].mean()

OR_TEMP     60.072583
OR_MIN      51.212050
OR_MAX      69.288994
OR_WDSP      7.691552
OR_MXSPD    17.057134
OR_GUST     26.882647
OR_PRCP      0.252017
OR_SNDP      5.227842
dtype: float64

At first glance the only numbers that indicate a large difference are precipitation which more than doubles and snow depth which increases by less about 15%. Other numbers that may indicate correlations are max windspeed and gust.